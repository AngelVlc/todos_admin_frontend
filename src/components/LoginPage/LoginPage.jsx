import React, { useContext, useState } from 'react';
import { useHistory } from "react-router-dom";
import { useFormik } from 'formik';
import { userLoggedIn } from '../../actions'
import { AppContext } from '../../contexts/AppContext'
import * as Yup from 'yup';
import jwt from 'jwt-decode';
import { doGetToken } from '../../helpers/api';


export const LoginPage = () => {
    const { authDispatch, requestsDispatch } = useContext(AppContext)
    const [authError, setAuthError] = useState(null);
    let history = useHistory();

    const formik = useFormik({
        initialValues: {
            userName: '',
            password: ''
        },
        validationSchema: Yup.object({
            userName: Yup.string()
                .required('Required'),
            password: Yup.string()
                .required('Required'),
        }),
        onSubmit: async (values) => {
            try {
                const res = await doGetToken(values, requestsDispatch);
                const tokenInfo = jwt(res.token);
                authDispatch(userLoggedIn(res, tokenInfo));
                history.push('/');
            } catch (error) {
                setAuthError(error);
            }
        },
    });

    return (
        <div>
            <h4>LOG IN PAGE</h4>
            <form onSubmit={formik.handleSubmit}>
                <label htmlFor="userName">Name</label>
                <input
                    id="userName"
                    name="userName"
                    onChange={formik.handleChange}
                    onBlur={formik.handleBlur}
                    value={formik.values.userName}
                />
                {formik.touched.userName && formik.errors.userName ? (
                    <div>{formik.errors.userName}</div>
                ) : null}
                <label htmlFor="password">Password</label>
                <input
                    id="password"
                    name="password"
                    type="password"
                    onChange={formik.handleChange}
                    onBlur={formik.handleBlur}
                    value={formik.values.password}
                />
                {formik.touched.password && formik.errors.password ? (
                    <div>{formik.errors.password}</div>
                ) : null}
                {authError ? (
                    <div>{authError}</div>
                ) : null}
                <button type="submit">Log In</button>
            </form>
        </div>
    );
}
