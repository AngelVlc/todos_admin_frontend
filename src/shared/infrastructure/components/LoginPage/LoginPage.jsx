import React, { useState, useContext, useEffect } from 'react';
import { AppContext } from '../../contexts';
import { useHistory } from "react-router-dom";
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { userLoggedIn, userLoggedOut } from '../../actions';
import * as Yup from 'yup';
import axios from 'axios';

export const LoginPage = () => {
    const { authDispatch } = useContext(AppContext);
    const [authError, setAuthError] = useState(null);
    let history = useHistory();

    useEffect(() => {
        authDispatch(userLoggedOut());
        localStorage.setItem('userInfo', null);
    }, [authDispatch]);

    const onSubmit = async(values) => {
        try {
            const res = await axios.post('/auth/login', values);
            localStorage.setItem('userInfo', JSON.stringify(res.data));
            authDispatch(userLoggedIn(res.data));
            history.push('/');
        } catch (error) {
            setAuthError(error);
        }
    }

    return (
        <div className="container">
            <h3 className="title">LOG IN</h3>
            <Formik
                initialValues={{
                    name: '',
                    password: ''
                }}
                validationSchema={Yup.object({
                    name: Yup.string().required('Required'),
                    password: Yup.string().required('Required'),
                })}
                onSubmit={onSubmit}>
                <Form>
                    <div className="field">
                        <label className="label" htmlFor="name">Name</label>
                        <div className="control">
                            <Field name="name" type="text" data-testid="name" autoFocus />
                        </div>
                        <p className="help is-danger" data-testid="nameErrors">
                            <ErrorMessage name="name" />
                        </p>
                    </div>

                    <div className="field">
                        <label className="label" htmlFor="password">Password</label>
                        <div className="control">
                            <Field name="password" type="password" data-testid="password" />
                        </div>
                        <p className="help is-danger" data-testid="passwordErrors">
                            <ErrorMessage name="password" />
                        </p>
                    </div>
                    <div className="control">
                        <button className="button" type="submit" data-testid="submit">Log In</button>
                        {authError ? (
                            <p className="help is-danger" data-testid="authError">{authError}</p>
                        ) : null}
                    </div>
                </Form>
            </Formik>
        </div>
    );
}
