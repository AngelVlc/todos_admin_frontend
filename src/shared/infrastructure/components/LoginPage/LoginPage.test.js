import { render, cleanup, fireEvent, waitFor } from '@testing-library/react'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import { LoginPage } from './LoginPage'
import { AppContext } from '../../contexts';
import axios from 'axios';

jest.mock('axios');

const mockAuthDispatch = jest.fn();
const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockHistoryPush
    })
}));

const renderWithContextAndRouter = () => {
    const history = createMemoryHistory();
    const context = { authDispatch: mockAuthDispatch };

    return {
        ...render(
            <AppContext.Provider value={context}>
                <Router history={history}>
                    {<LoginPage />}
                </Router>
            </AppContext.Provider>)
    }
}

beforeEach(() => {
    Object.defineProperty(window, "localStorage", {
        value: {
          getItem: jest.fn(() => null),
          setItem: jest.fn(() => null)
        },
        writable: true
    });
});

afterEach(cleanup);

it('should match the snapshot', () => {
    const { asFragment } = renderWithContextAndRouter();

    expect(asFragment()).toMatchSnapshot();
});

it('should require user name and password for logging in', async () => {
    const { getByTestId } = renderWithContextAndRouter();

    await waitFor(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(getByTestId('userNameErrors')).toHaveTextContent('Required');
    expect(getByTestId('passwordErrors')).toHaveTextContent('Required');
})

it('should show the error when log in fails', async () => {
    const { getByTestId } = renderWithContextAndRouter();

    await changeInput(getByTestId, 'userName', 'user');
    await changeInput(getByTestId, 'password', 'pass');

    axios.post.mockRejectedValue('some error');

    await waitFor(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(axios.post.mock.calls.length).toBe(1);
    expect(axios.post.mock.calls[0][0]).toBe('/auth/login');
    expect(axios.post.mock.calls[0][1]).toStrictEqual({userName: 'user', password: 'pass'});
    expect(getByTestId('authError')).toHaveTextContent('some error');
    axios.post.mockClear();
})

it('should show the home page after logging in', async () => {
    const { getByTestId } = renderWithContextAndRouter();

    await changeInput(getByTestId, 'userName', 'user');
    await changeInput(getByTestId, 'password', 'pass');

    axios.post.mockResolvedValue({
        data: {
            userId: 1,
            userName: 'userName',
            isAdmin: true
        }
    });

    await waitFor(() => {
        fireEvent.click(getByTestId('submit'));
    });

    expect(axios.post.mock.calls.length).toBe(1);
    expect(axios.post.mock.calls[0][0]).toBe('/auth/login');
    expect(axios.post.mock.calls[0][1]).toStrictEqual({ password: "pass", userName: "user" });
    expect(window.localStorage.setItem.mock.calls.length).toBe(2);
    expect(window.localStorage.setItem.mock.calls[0][0]).toBe('userInfo');
    expect(window.localStorage.setItem.mock.calls[0][1]).toBe(null);
    expect(window.localStorage.setItem.mock.calls[1][0]).toBe('userInfo');
    expect(window.localStorage.setItem.mock.calls[1][1]).toBe(JSON.stringify({ userId: 1, userName: 'userName', isAdmin: true}));
    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe('/');
    axios.post.mockClear();
})

const changeInput = async (getByTestId, name, value) => {
    await waitFor(() => {
        fireEvent.change(getByTestId(name), {
            target: {
                value: value
            }
        })
    })
};
