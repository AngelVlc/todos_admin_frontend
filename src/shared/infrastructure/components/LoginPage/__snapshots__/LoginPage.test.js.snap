// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      LOG IN
    </h3>
    <form
      action="#"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="username"
        >
          Name
        </label>
        <div
          class="control"
        >
          <input
            data-testid="name"
            name="username"
            type="text"
            value=""
          />
        </div>
        <p
          class="help is-danger"
          data-testid="nameErrors"
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="password"
        >
          Password
        </label>
        <div
          class="control"
        >
          <input
            data-testid="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <p
          class="help is-danger"
          data-testid="passwordErrors"
        />
      </div>
      <div
        class="control"
      >
        <button
          class="button"
          data-testid="submit"
          type="submit"
        >
          Log In
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
